File: java/src/com/company/datastructureimpl/Trie.java
package com.company.datastructureimpl;

import java.util.HashMap;
import java.util.Map;

// trie node
class TNode {
    public final Map<Character, TNode> children =
            new HashMap<>();
    public boolean endOfString = false;
}

public class Trie {

    private final TNode root = new TNode();

    // time O(L), space O(1)
    public boolean exists(final String word) {
        var cur = root;
        for (var letter : word.toCharArray()) {
            if (!cur.children.containsKey(letter)) return false;
            cur = cur.children.get(letter);
        }
        return cur.endOfString;
    }

    // time O(L), space O(L)
    public void insert(final String word) {
        var cur = root;
        for (int i = 0; i < word.length(); i++) {

            var letter = word.charAt(i);

            if (!cur.children.containsKey(letter)) cur.children.put(letter, new TNode());

            cur = cur.children.get(letter);

        }
        cur.endOfString = true;
    }

    public static void main(String[] args) {
        var trie = new Trie();
        trie.insert("app");
        trie.insert("api");
        trie.insert("apis");
        trie.insert("s");

        assert trie.exists("app");
        assert trie.exists("api");
        assert trie.exists("apis");
        assert trie.exists("s");
        assert !trie.exists("ap");
        assert !trie.exists("apiss");
        assert !trie.exists("a");
        assert !trie.exists("");

        System.out.println("tests passed");
    }
}

File: python/MIT_course/trie.py
class Trie:

    def default_dict(self):
        return {"end": False}

    def __init__(self):
        self.root = self.default_dict()

    def insert(self, word: str) -> None:
        cur = self.root
        for letter in word:
            if letter not in cur:
                cur[letter] = self.default_dict()

            cur = cur[letter]
        cur["end"] = True

    def exists(self, word: str) -> bool:
        cur = self.root
        for letter in word:
            if letter not in cur:
                return False
            cur = cur[letter]
        return cur["end"]


trie = Trie()
trie.insert("app")
trie.insert("apps")
trie.insert("banana")
trie.insert("ban")

assert trie.exists("app")
assert trie.exists("apps")
assert trie.exists("banana")
assert trie.exists("ban")
assert not trie.exists("ba")
assert not trie.exists("ap")
assert not trie.exists("appps")

print("tests passed")
